Sandbox function

sisp opened the pull request:
In continuation of our discussion in #73, this is a first attempt at creating a middleware for managing a draft copy of a tree for efficient testing of changes that can be applied to the original tree or rejected.

In most cases, the middleware must be applied to the root of a tree, e.g. in order to support `rootRef` resolution in the draft as well. When the middleware is applied to a root store, the draft is also registered as a root store.

In order to access `withDraft` anywhere in a tree, I would provide `DraftManager` using a context.

What do you think?

xaviergonz commented on packages/lib/src/actionMiddlewares/draftMiddleware.ts:
```diff
@@ -0,0 +1,102 @@
+   */
+  public constructor(private readonly subtreeRoot: object) {
+    assertTweakedObject(subtreeRoot, "subtreeRoot")
+    this.subtreeRootDraft = clone(subtreeRoot)
```
pass `{generateNewIds: false}` or else the $modelIds of the clone will be different from the original

sisp replied:
Good point! Done in bb70a5b73446de0e62036699198461f25911a390.

xaviergonz commented on packages/lib/src/actionMiddlewares/draftMiddleware.ts:
```diff
@@ -0,0 +1,102 @@
+  public constructor(private readonly subtreeRoot: object) {
+    assertTweakedObject(subtreeRoot, "subtreeRoot")
+    this.subtreeRootDraft = clone(subtreeRoot)
+    if (isRootStore(subtreeRoot)) {
```
Just made `isRootStore` reactive in 0.29.0, so it might be worth to do this like
```
let wasRS = false;
reaction(() => isRootStore(subTreeRoot), (isRS) => {
  if (isRS !== wasRS) {
    wasRS = isRS
    if (isRS) { registerRootStore(this.subtreeRoofDraft) }
    else { unregisterRootStore(this.subtreeRoofDraft) }
 }
  }
}, {fireImmediately: true})
```
That way the draft will follow the state of the original one

sisp replied:
That's a great idea! Done in 7b2cad4dfd9418dcd72347359ecc0b2ecfe1ea2a.

Do you like the way I'm dealing with the disposers? I was thinking about using an array of disposers (the way it's done now) or assigning a closure with the disposers to the private `disposer` class property. By using the array, I can remove the disposers from the array once they have been called - not sure if that's really necessary. In general, what behavior do you expect when a disposer is called multiple times?

xaviergonz replied:
It should do nothing when called more than once, so that approach seems OK to me

xaviergonz commented on packages/lib/src/actionMiddlewares/draftMiddleware.ts:
```diff
@@ -0,0 +1,102 @@
+   * applied to the original subtree when `fn` returns `true`. When `fn` returns `false`,
+   * `undefined` or nothing, the changes made to the draft are rejected.
+   */
+  public withDraft<T extends object>(node: T, fn: WithDraftCallback<T>): void {
```
just wondering, if there any reason to take the selection of a node rather than just pass the whole object and do the selection inside the draft function if needed?

I don't think there would be any kind of speed penalty

Also that way it could make typing less error prone (DraftManager could have type T, which is passed from `draftManager` function)

xaviergonz replied:
Basically what I mean is that it could be done as 
```
const dm = draftManager(something) // T is typeof something
dm.withDraft(draft => { // draft is WithDraftCallback of T above
  const selector = draft.a.b.c,
})
```

rather than
```
const dm = draftManager(something)
dm.withDraft(something.a.b.c, selector => {
})
```

and that way you automatically ensure you can't pass invalid children as first parameter to the function (since there´s no such first parameter).

xaviergonz replied:
Ahh, but maybe you want it to make it easy to select a node on the draft given the original. If so nevermind :)

sisp replied:
> Ahh, but maybe you want it to make it easy to select a node on the draft given the original. If so nevermind :)

Yes, exactly. I'd provide the `DraftManager` instance to the children of the subtree root and select the draft of an original node from, e.g., a computed property or `onAttachedToRootStore`.

xaviergonz commented on packages/lib/src/actionMiddlewares/draftMiddleware.ts:
```diff
@@ -0,0 +1,102 @@
+    if (accept) {
+      applyPatches(
+        this.subtreeRoot,
+        recorder.events.flatMap(event => event.patches)
```
Just wondering, how widespread is the support for flatmap nowadays?

sisp replied:
https://caniuse.com/#feat=array-flat

IE and Edge don't seem to support it plus a few other (mobile) browsers. After checking `mobx-keystone`'s other requirements like `WeakMap` etc., I think `flatMap` would reduce compatibility, so I'll use an alternative.

sisp replied:
Done in 7c67c7084d9b812f4468b441ee704ee6d63436c5. Do you prefer calling `applyPatches` multiple times (like it's done now) or materializing a new array and calling `applyPatches` only once?

There was a bug with regard to inverse patches order which I fixed in 91743704ebce5173fb7a6954adc2219cccc2df90.

xaviergonz replied:
For performance reasons I usually prefer to avoid creating array copies whenever possible, so this approach looks good.

xaviergonz commented on packages/lib/src/actionMiddlewares/draftMiddleware.ts:
```diff
@@ -0,0 +1,102 @@
+   */
+  public dispose(): void {
+    this.disposer()
+    if (isRootStore(this.subtreeRootDraft)) {
```
remember to dispose of the reaction here if done in a reactive way (as suggested above) while keeping this code too

sisp replied:
Done, but also see my above comment about the approach I've taken in general.

xaviergonz replied:
LGTM

sisp replied:
Done in d21afe4e27488cef866e624c117c67f53db52444. The disposer closure becomes a noop after it has been called once.

xaviergonz commented on packages/lib/src/actionMiddlewares/draftMiddleware.ts:
```diff
@@ -0,0 +1,102 @@
+
+    const path = getParentToChildPath(this.subtreeRoot, node)
+    if (!path) {
+      return
```
maybe `throw failure("...")`

sisp replied:
Done in c8a1f257c4ea68ab2a7540bcdefe0cd2dc34e9a9.

xaviergonz commented on packages/lib/src/actionMiddlewares/draftMiddleware.ts:
```diff
@@ -0,0 +1,102 @@
+
+    const draftNode = resolvePath<T>(this.subtreeRootDraft, path).value
+    if (!draftNode) {
+      return
```
maybe `throw failure("...")`

sisp replied:
After looking at this again, I think an error check is not needed at all because the path must be resolvable. Do you agree?

xaviergonz replied:
Wouldn't it fail if for some reason the draft is out of sync with the original?

sisp replied:
You're right. Ideally, this should never happen, but I've added the check in e79e4656b26297cea09a316b6a8d74e706e64a70 to be on the safe side.

xaviergonz commented on packages/lib/src/actionMiddlewares/draftMiddleware.ts:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    const recorder = patchRecorder(this.subtreeRootDraft)
+    const accept = fn(draftNode)
```
try/finally and in the finally block call the disposer (to avoid mem leaks) just in case the userland accept fn throws

xaviergonz replied:
and maybe run the "non-accept" code in case it throws (to keep the draft consistent with the original in such cases)

sisp replied:
If the userland accept fn throws and I call the disposer in the finally block but the user catches the exception and recovers properly, won't calling the disposer cause an unexpected state of the `DraftManager` instance? I'd say the user is responsible for disposing of the `DraftManager` instance manually.

xaviergonz replied:
Sorry, I wasn't specific enough, I meant the recorder disposer 😁

sisp replied:
Of course, that makes sense. Done in 39e4911f81f180d04b1e727953998bc98cb5f92d.

I decided to handle both `accept` cases in the finally block as I think it's a bit easier to read instead of handling only falsy `accept` in the finally block and truthy `accept` after the finally block. Would you prefer handling only falsy `accept` in the finally block?

xaviergonz replied:
Looks nice!

sisp commented:
I've completed addressing your comments above and replied to them individually. Let me know what you think. :-)

xaviergonz commented on packages/lib/src/actionMiddlewares/draftMiddleware.ts:
```diff
@@ -0,0 +1,123 @@
+    } finally {
+      recorder.dispose()
+      if (accept) {
+        for (const event of recorder.events) {
```
looks good, but bear in mind that the transpilation of for-of is kind of slow. if you want it to be more performant then use something like
```
const len = arr.length
for (let i = 0; i < len; i++) {....}
```

sisp replied:
Uh, I didn't know about the transpilation to slow code. I changed it as you suggested in 78c2d3dd0b66100c57ff3bcbaecf79f3927a02a4.

xaviergonz commented on packages/lib/src/actionMiddlewares/draftMiddleware.ts:
```diff
@@ -0,0 +1,123 @@
+   * are applied to the original subtree when `fn` returns `true`. When `fn` returns `false`,
+   * `undefined` or nothing, the changes made to the draft are rejected.
+   */
+  public withDraft<T extends object>(node: T, fn: WithDraftCallback<T>): void {
```
it just occurred to me that it might be worth to decorate this function with mobx action to ensure all done inside this function (e.g. applyPatches) (and inside the fn function) is a single transaction and is untracked

sisp replied:
If the `withDraft` method was decorated with `@action`, the result of a reaction triggered by a change in `fn` would not be available in `fn`, would it? Because the reaction would be executed after `fn` has returned. Then, I wouldn't be able use the result of a change made to the draft in order to determine whether the change should also be applied to the original tree or rejected.

I could run the loop around `applyPatches` in an adhoc action to at least apply (inverse) patches in a single transaction. As for `fn`, I think the user should remain in control what part of `fn` should be executed as an action. I'd add a note about this in the documentation that I'll write once we've settled on the implementation.

xaviergonz replied:
Good, sounds like a good compromise :)

sisp replied:
Done in a70cdd3b448290798799445367d2e1251ca13509.

sisp replied:
Should I write a test that multiple patches are applied in a single transaction?

xaviergonz replied:
No need I think, that would be testing mobx functionality, wouldn't it?

sisp replied:
Well, we know that `applyPatches` is now wrapped in `runInAction` and that this function is tested, but none of the current tests checks the behavior of applying patches in a single transaction. Do you know what I mean?

xaviergonz replied:
What would the test be like, checking if a reaction triggers only once and only after the the multiple patches wrapped inside a single runAction are run?

If so, I see it no different from doing any other kind of mutation (mobx side test).

Maybe I'm missing something though... 😕

sisp replied:
> What would the test be like, checking if a reaction triggers only once and only after the the multiple patches wrapped inside a single runAction are run?

Yes, that would have been my suggestion.

> If so, I see it no different from doing any other kind of mutation (mobx side test).
>
> Maybe I'm missing something though... :confused:

Hm, I guess you're right, no test. :-)

xaviergonz commented:
Looks better! Just added a few more comments, hope you don't mind those.

sisp commented:
> Looks better! Just added a few more comments, hope you don't mind those.

I appreciate your thorough review, no worries! :-)

I'm done with the next iteration except the open discussion about decorating `withDraft` or `fn` with `@action`.

xaviergonz commented:
Approved :) Feel free to merge when you feel ready

sisp commented:
Thanks for approving the PR and for your valuable feedback and advice! Two questions left:

1. Would you like me to squash the commits to keep the history clean of the iterations needed to get to this point?
1. I'd be happy to write documentation. Would you like it to be part of this PR or rather a new one?

xaviergonz commented:
> Thanks for approving the PR and for your valuable feedback and advice! Two questions left:

Thanks to you for creating it!

> 1. Would you like me to squash the commits to keep the history clean of the iterations needed to get to this point?

I usually squash it, but I really have no strong preference.

> 2. I'd be happy to write documentation. Would you like it to be part of this PR or rather a new one?

Whichever you prefer is fine by me :)

sisp commented:
I've added a documentation page for the draft middleware. I didn't find it easy to come up with a simple example to convey the gist of this middleware. As a result, in my opinion the overview example is simple but doesn't demonstrate the need for making changes to a draft as the problem could easily be solved "conventionally" without a draft. In the second example, I hope to convey the merit a bit better at the cost of more code and text.

Your feedback is again much appreciated. :-)

xaviergonz commented:
The second example got me a bit lost TBH. What about a more usual use case such as a form that can "select" part of the root store to edit it (for example a profile) and then can either be committed (Save) or not (Cancel)?

sisp commented:
> The second example got me a bit lost TBH.

Okay, that's not good. I was trying to come up with an example that cannot be done easily without the draft middleware. In many simple examples (at least what I have in mind right now), it is possible to implement a model like this:

```ts
class MyModel extends Model({ value: prop<...>() }) {
  validate(value: ...): string | undefined {
    // return an error message or `undefined`
  }

  @computed
  get error(): string | undefined {
    return this.validate(this.value)
  }
}
``` 

In order to check whether a new value would be valid without actually mutating the state, `validate` can be called with the new value. But in more complex cases whose validation may depend on information beyond the model instance itself, or especially when the results of side effects must be taken into account, a draft makes validation much easier as validation logic can be implemented straightforward against the current state of the tree.

> What about a more usual use case such as a form that can "select" part of the root store to edit it (for example a profile) and then can either be committed (Save) or not (Cancel)?

I'm afraid I'm not quite sure what you mean exactly. Something like this?

```ts
const draftCtx = createContext<DraftManager>()

@model("MyApp/Profile")
class Profile extends Model({
  username: prop<string>(),
}) {
  @computed
  get errors(): { username?: string } {
    return this.username.length < 3 || this.username.length > 20
      ? { username: "Must have between 3 and 20 characters" }
      : undefined
  }

  @modelAction
  setUsername(username: string): void {
    this.username = username
  }
  
  validateUsername(username: string): boolean {
    let isValid = false
    draftCtx.get(this)?.withDraft(this, draft => {
      draft.setUsername(username)
      isValid = this.errors.username === undefined
    })
    return isValid
  }
}

const profile = new Profile({ username: "sisp" })
const profileDraftManager = draftMiddleware(profile)
draftCtx.setComputed(profile, () => profileDraftManager)

profile.errors // => {}
profile.validateUsername("xaviergonz") // => `undefined`
profile.validateUsername("u") // => "Must have between 3 and 20 characters"
```

xaviergonz commented:
Nevermind, I was actually thinking of something along the lines of createViewModel from mobx-utils

https://github.com/mobxjs/mobx-utils/blob/master/src/create-view-model.ts

Anyway don't worry too much about the examples, it can be improved at any time in the future.

Btw, do you think it makes sense to add support for async functions to withDraft? (or adding a separate withDraftAsync)

sisp commented:
> Nevermind, I was actually thinking of something along the lines of createViewModel from mobx-utils
>
> https://github.com/mobxjs/mobx-utils/blob/master/src/create-view-model.ts

Although the concepts are similar, I'm not sure what the equivalent example of `createViewModel` would be for `withDraft` because the lifecycle of the changes made to the view model is handled manually whereas `withDraft` manages the lifecycle for you.

> Btw, do you think it makes sense to add support for async functions to withDraft? (or adding a separate withDraftAsync)

I think it's a good idea support async functions, too. I'll try come up with something and get back to you.

sisp commented:
After giving it some thought, I'm a bit stuck with supporting async functions.

Currently, `withDraft` manages the state of the draft synchronously, i.e. changes made to the draft in `fn` are either applied to the original subtree or the draft is reset to its previous state when `fn` returns.

What would be the signature of, let's say, `withDraftAsync`? I assume:

```ts
withDraftAsync<T extends object>(node: T, fn: (draft: T) => Promise<boolean | undefined | void>): void
// or Promise<void> to make `withDraftAsync` async itself?
```

Assume the following call:

```ts
draftManager.withDraftAsync(node, async draft => {
  draft.something()
  await someAsyncFn()
  draft.somethingElse()
})
```

I think this call roughly corresponds to (correct me if I'm wrong):

```ts
draftManager.withDraftAsync(node, draft => {
  draft.something()
  Promise((resolve, reject) => {
    resolve(someAsyncFn())
  }).then(() => {
    draft.somethingElse()
  })
})
```

But the callback passed to `then` is a closure which accesses `draft` from the outer scope and since the callback is executed at a later point, the state of `draft` is not valid anymore when the callback is executed. Do you see what I mean? If done this way, somehow `draft` would need to be restored to the correct state before the callback is executed.

I've noticed that `modelFlow` is implemented using generators. Would using generators help and be the right thing to do here?

Sorry, I still lack some experience with asynchronous programming in JS.

sisp commented:
I may be able to build something like this:

```ts
draftManager.withDraftAsync(node, async withDraft => {
  withDraft(draft => {
    draft.something()
  })
  await someAsyncFn() // or, e.g., `await someAsyncFn(withDraft)`
  withDraft(draft => {
    draft.somethingElse()
  })
})
```

`withDraft` is a synchronous function which receives a synchronous callback with `draft` as an argument. `withDraft` is responsible for applying the sequence of patches up to that point to pass the correct state of `draft`  to the callback and applies the inverse patches afterwards to reset `draft` again. Instead of accessing `draft` in a closure, `withDraft` is accessed (or passed around).

The downside of this approach is that patches and inverse patches may need to be applied multiple times, so there is some overhead.

xaviergonz commented:
Maybe something like this? https://github.com/sisp/mobx-keystone/pull/2/commits/c918e03031aa3456c5c9ae23a34c7fdaeb26d416

sisp commented:
Did I just overwrite your commit? :thinking: I force-pushed because I had fixed a comment typo in the first commit.

sisp commented:
> So, if I get it correctly:
>
> ```ts
> draftManager.withDraftAsync(node, async withDraft => {
>   // this block is recorded
>   withDraft(draft => {
>     draft.something()
>   })
> 
>   // no recording here
>   await someAsyncFn() // or, e.g., `await someAsyncFn(withDraft)`
> 
>   // this block is recorded
>   withDraft(draft => {
>     draft.somethingElse()
>   })
> 
>   // <-- here the user needs to return true or false to accept/skip all previous changes?
> })
> ```

Yes, every `withDraft` within `withDraftAsync` continues recording changes.

> Anyway, if you want it might be better to just skip async for now and go with the original version (plus the lock + maybe force withDraft to return a boolean). That should be good enough for a first version.

I'd be fine with that and continue working on async in a separate PR. I think the lock is a good idea. Would you like me to copy/include the lock part from your commit (that I may have accidentally overwritten - very sorry about that) or do you want to commit these changes?

I'd be fine with returning a boolean, so the intent is clear. Before making this change, I'd like to briefly discuss an alternative API that I started thinking about:

```ts
type WithDraftCallback<T extends object, R> = (draftNode: T, accept: () => void) => R

withDraft<T extends object, R = void>(node: T, fn: WithDraftCallback<T, R>): R {
  // ...
}
```

With this API, `withDraft` can return a value (or object or whatever) that is computed in `fn` based on the draft. A user would write the changes made to the draft by calling `accept()` (or it could also be named `commit` or `submit`), typically right before the return statement. E.g.:

```ts
const draftError = draftManager.withDraft(node, draft => {
  draft.setValue(...)
  return draft.error
})
```

Whereas with the current API:

```ts
let draftError
draftManager.withDraft(node, draft => {
  draft.setValue(...)
  draftError = draft.error
  return false // currently optional but could be required as you suggested
})
```

When making changes to the draft that should be applied to the original subtree:

```ts
draftManager.withDraft(node, (draft, accept) => {
  draft.setValue(...)
  if (!draft.error) {
    accept()
  }
  return ... // If you want to return something
})
```

What do you think?

xaviergonz commented:
Please copy the lock, the PR was just as an idea :)

IMHO submit() as a parameter could make it very easy to forget to call it (difficult to enforce), where as with "return boolean" TS will nag you about that boolean in compilation time (and it might even throw in runtime if a boolean is not returned as a fail-safe).

What could be done however is support two kind of return values, the boolean or a
```
{
  commit: boolean; // or submit, or accept, etc.
  return: R // or returnValue, etc.
}
```
or something similar and infer the return type from there (void if a boolean, R if that kind of object).

xaviergonz commented:
Typings:

```ts
type WithDraftCallback<T extends object, R > = (node: T) => boolean | {commit: boolean; return: R}

function test<T extends object, R = void>(node: T, cb: WithDraftCallback<T, R>): R {
  return null as any
}

// a is void
const a = test({x: 5}, (n) => {
  return true
})

// b is number
const b = test({x: 5}, (n) => {
  return {commit: true, return: 5}
})
```

sisp commented:
I have

- removed the commit for `withDraftAsync`,
- added the lock to prevent concurrent calls of `withDraft`, and
- added support to `withDraft` for returning a value from `fn`.

Meanwhile, I noticed that it was possible to get a reference of the draft and mutate the draft outside `fn` which should not be allowed as the draft and the original subtree would get out of sync. I'm now using the readonly middleware to protect the draft. The only thing that I don't quite understand is why

```ts
const disposeOnPatches = onPatches(subtreeRoot, patches =>
  applyPatches(this.subtreeRootDraft, patches)
)
```

works without wrapping `applyPatches` with `allowWrite` whereas `allowWrite` is needed when reverting changes made to the draft:

```ts
this.allowWrite(() => {
  let i = recorder.events.length
  while (i--) {
    applyPatches(this.subtreeRootDraft, recorder.events[i].inversePatches)
  }
})
```

Do you know the reason?

sisp commented:
> But now that I think about it, withDraft cannot (shouldn't) be run concurrently, if done you'd have multiple patches recorded for the same actions (I added a lock to the commit above to avoid that).
>
> _From:_ https://github.com/xaviergonz/mobx-keystone/pull/77#issuecomment-569420128

Now that I think about it, `withDraft` shouldn't be run _concurrently_ (in the async sense), but I think it _should_ be possible to nest `withDraft` calls. Think about model actions that use `withDraft` and one model action that uses `withDraft` may call another action that also uses `withDraft`. This scenario seems quite realistic and wouldn't be supported with the lock.

sisp commented:
I have a working implementation of nested `withDraft` calls, but it needs some cleanup. I'll push a commit tomorrow for you to review.

xaviergonz commented:
> Meanwhile, I noticed that it was possible to get a reference of the draft and mutate the draft outside fn which should not be allowed as the draft and the original subtree would get out of sync. I'm now using the readonly middleware to protect the draft. The only thing that I don't quite understand is why

Thanks, that will be fixed on 0.30.1

> I have a working implementation of nested withDraft calls, but it needs some cleanup. I'll push a commit tomorrow for you to review.

👍 nice!

sisp commented:
I've pushed a commit that adds support for nested `withDraft` calls. Essentially, there is only one patch recorder, but each `withDraft` call checks how many recorder events have been recorded prior to this `withDraft` call. Any additional events added to the recorder belong to this (nested) `withDraft` call or another nested one. When `fn` returns, committing changes to a draft (when `withDraft` calls are nested) means doing nothing because the changes are part of the recorder's recorded events already. When we're back at the outermost `withDraft` call, patches are applied to the original subtree as before. When changes are to be rejected, inverse patches are applied to the draft as before _but_ only the ones that belong to the nested `withDraft` call.

Let me know what you think. 🙂

> Thanks, that will be fixed on 0.30.1

Great, thanks! I've rebased the branch to include this fix and added the missing `allowWrite`.

xaviergonz commented on packages/lib/src/actionMiddlewares/draftMiddleware.ts:
```diff
@@ -0,0 +1,196 @@
+  ): { draftNode: T; applyRecorderChanges: (commit: boolean) => void } {
+    assertTweakedObject(node, "node")
+
+    const isDraft = !!this.withDraftPatchRecorder
```
I assume isDraft means "isNestedWithDraftCall" ?

sisp replied:
Yes, you're right, I'll change the name.

sisp replied:
Done

xaviergonz commented on packages/lib/src/actionMiddlewares/draftMiddleware.ts:
```diff
@@ -0,0 +1,196 @@
+      this.withDraftPatchRecorder = patchRecorder(this.subtreeRootDraft)
+    }
+    const recorder = this.withDraftPatchRecorder
+    const numRecorderEvents = recorder.events.length
```
ah, so you only rollback the amount of events that the inner call generated, smart!

sisp replied:
Yes, thanks 🙂

xaviergonz commented on packages/lib/src/actionMiddlewares/draftMiddleware.ts:
```diff
@@ -0,0 +1,196 @@
+      }
+      runInAction(() => {
+        if (commit) {
+          if (!isDraft) {
```
I'm torn about this.
On the one hand it might be confusing that the inner call does not actually commit anything, but on the other hand it would be weird for the inner call to commit changes it did not make.
Maybe the inner draft should internally generate another draft object (draft of the draft?), but that would be slow...

What are your thoughts on this?

xaviergonz replied:
Ah, and I think this behavior (whichever is chosen) should be documented.

sisp replied:
> Maybe the inner draft should internally generate another draft object (draft of the draft?), but that would be slow...

Yes, ideally there should be no need to have a centrally managed draft, but cloning often is too slow. ...

> On the one hand it might be confusing that the inner call does not actually commit anything, but on the other hand it would be weird for the inner call to commit changes it did not make.

... So in my opinion this is an implementation detail that is necessary to support this API/DX. A developer shouldn't notice that committing changes from an inner call doesn't actually do anything because the changes have already been recorded due to the way this mechanism is implemented.

Or are you concerned about a userland `onPatches` listener that gets triggered slightly unexpectedly (not when the nested `withDraft` returns but immediately when the changes are made)?

xaviergonz replied:
My concern is the duality when returning true:
- if it is un-nested it will commit changes right away
- if it is nested it will commit changes only if (and after) the outermost withDraft call returns true

I have the hunch that is the right behavior, but I also think it should be on the docs so it doesn't surprise anybody :)

sisp replied:
Okay, I'll add documentation about this behavior.

xaviergonz commented:
All looks good to me btw :)

sisp commented:
Do you think this middleware satisfies the properties of a middleware? To me a middleware enhances the tree in some way, e.g. track actions or make the tree readonly. I'm not sure how this feature enhances the tree. It rather creates an object that manages a draft copy and provides an API (`withDraft`) for making changes to the draft with commit/reject action managed by the lifecycle of the `withDraft` call.

xaviergonz commented:
Yeah, guess it is more an utility than a middleware. Also it feels more like a "sandbox" (like you said, for "what-if" scenarios) utility than a "draft". So maybe it would be better to call it Sandbox (Playground or something similar) and put it into its own section like "Drafts" now are?

sisp commented:
Sounds good, I'll rename it and make it a utility instead of a middleware. Thanks! :-)

sisp commented:
I've renamed this utility to "sandbox" and tried to improve documentation.

xaviergonz commented:
Excellent! Just wondering, why do nested calls need to use a sandbox node?
Is it ready to be merged then btw?

sisp commented:
I was thinking that someone might, e.g., detach/remove a model instance in the outer `withSandbox` call and then try to get the sandbox node of this detached/removed model instance in the inner `withSandbox` call using the node from the original tree which still exists there but doesn't exist in the sandbox anymore.

Instead of disallowing nested `withSandbox` calls with a node from the original subtree, I could also do this:

```diff
--- a/packages/lib/src/treeUtils/sandbox.ts
+++ b/packages/lib/src/treeUtils/sandbox.ts
@@ -138,10 +138,9 @@ export class SandboxManager {
 
     const isNestedWithSandboxCall = !!this.withSandboxPatchRecorder
 
-    const path = getParentToChildPath(
-      isNestedWithSandboxCall ? this.subtreeRootClone : this.subtreeRoot,
-      node
-    )
+    const path =
+      getParentToChildPath(this.subtreeRoot, node) ||
+      getParentToChildPath(this.subtreeRootClone, node)
     if (!path) {
       throw failure(`node is not a child of subtreeRoot${isNestedWithSandboxCall ? "Clone" : ""}`)
     }
```

> Is it ready to be merged then btw?

Except the point you raised above, I think so. Before merging this PR, I'd squash the commits again to clean up the history and then you can merge it. But I think we should resolve your point about requiring a sandbox node in nested `withSandbox` calls first.

sisp commented:
I've squashed the commits and added an entry in the changelog. Feel free to merge this PR if you think there's nothing more to discuss about nested `withSandbox` calls requiring a sandbox node, but I'm also happy to continue the discussion if you are in doubt about this requirement. :-)

xaviergonz commented:
Sorry about the delay. Merged, I don't think many people would use nested calls anyway. Thanks for the contribution! I will release a new version today

sisp commented:
Thanks for all your feedback and support! :-)
